### PR TITLE
Fix CustomShuffleReader replacement when decimal types enabled

### PR DESCRIPTION
**Pull request by sperlingxx**
This pull request is to fix the bug of AQE with decimal, which is reported in issue #1670.

### CODE REVIEW

**Comment by sperlingxx** [comment]
build

File: tests/src/test/scala/com/nvidia/spark/rapids/AdaptiveQueryExecSuite.scala [hunk 1]
```diff
@@ -31,6 +31,7 @@ import org.apache.spark.sql.execution.joins.SortMergeJoinExec
 import org.apache.spark.sql.functions.{col, when}
```
Comment (jlowe) [inline tests/src/test/scala/com/nvidia/spark/rapids/AdaptiveQueryExecSuite.scala:31]: 2021 copyrights

**Review by jlowe — COMMENTED** [review]
Copyrights need to be updated but otherwise lgtm.

**Comment by revans2** [comment]
@sperlingxx Can this impact other types too?  We support shuffling nested types like arrays, maps, and structs, but GpuCustomShuffleReader does not.

**Comment by andygrove** [comment]
Thanks @sperlingxx . I applied this change to my branch and ran quite a few queries with decimal + AQE enabled and it worked well.

**Comment by abellina** [comment]
Is there a way to make this fail in a way that we can fallback to the cpu?

File: tests/src/test/scala/com/nvidia/spark/rapids/AdaptiveQueryExecSuite.scala (hunk 1 above)
Comment (sperlingxx) [reply tests/src/test/scala/com/nvidia/spark/rapids/AdaptiveQueryExecSuite.scala:31]: Fixed.

**Comment by sperlingxx** [comment]
> @sperlingxx Can this impact other types too? We support shuffling nested types like arrays, maps, and structs, but GpuCustomShuffleReader does not.

Hi @revans2, I added ArrayType and StructType as supported types. And MapType is not supported in `ShuffleExchangeExec`.

**Comment by sperlingxx** [comment]
> Is there a way to make this fail in a way that we can fallback to the cpu?

I believe when we configure `spark.rapids.sql.decimalType.enabled` with false, it will fallback to cpu. The fallback strategy will fail at here because parent plan(CustomShuffleReaderExec) fallbacks but child plan does not.

**Comment by sperlingxx** [comment]
build

File: docs/supported_ops.md [hunk 2]
```diff
@@ -9,7 +9,7 @@ support all data types. The RAPIDS Accelerator for Apache Spark has further
 restrictions on what types are supported for processing. This tries
 to document what operations are supported and what data types each operation supports.
 Because Apache Spark is under active development too and this document was generated
-against version 3.0.0 of Spark. Most of this should still
+against version 3.0.1 of Spark. Most of this should still
```
Comment (revans2) [inline docs/supported_ops.md:12]: The default is still to use 3.0.0 for building the docs.  What happened to change it to 3.0.1 when the build itself did not change?

Comment (sperlingxx) [reply docs/supported_ops.md]: My bad. I ran `mvn verify` with `Dspark.version=3.0.1` to update the doc.

**Comment by sperlingxx** [comment]
build

**Review by jlowe — APPROVED** [review]

**Pull request merged**